modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C

```diff
@@ -0,0 +1,590 @@
+
+    MooseVariable * rot_variable = getVar("rotations", i);
+    _rot_num[i] = rot_variable->number();
+  }
+
+  if (_large_strain && (_Ay[0] > 0.0 || _Ay[1] > 0.0 || _Az[0] > 0.0 || _Az[1] > 0.0))
+    mooseError("ComputeIncrementalBeamStrain: Large strain calculation does not currently "
+               "support asymmetric beam configurations with non-zero first or third moments of "
+               "area.");
+
+  for (unsigned int i = 0; i < _eigenstrain_names.size(); ++i)
+  {
+    _disp_eigenstrain[i] = &getMaterialProperty<RealVectorValue>("disp_" + _eigenstrain_names[i]);
+    _rot_eigenstrain[i] = &getMaterialProperty<RealVectorValue>("rot_" + _eigenstrain_names[i]);
+    _disp_eigenstrain_old[i] =
+        &getMaterialPropertyOld<RealVectorValue>("disp_" + _eigenstrain_names[i]);
+    _rot_eigenstrain_old[i] =
+        &getMaterialPropertyOld<RealVectorValue>("rot_" + _eigenstrain_names[i]);
+  }
+}
+
+void
+ComputeIncrementalBeamStrain::initQpStatefulProperties()
+{
+  // compute initial orientation of the beam for calculating initial rotation matrix
+  const std::vector<RealGradient> * orientation =
+      &_subproblem.assembly(_tid).getFE(FEType(), 1)->get_dxyzdxi();
+  RealGradient x_orientation = (*orientation)[0];
+  x_orientation /= x_orientation.norm();
+
+  RealGradient y_orientation = getParam<RealGradient>("y_orientation");
+  y_orientation /= y_orientation.norm();
+  Real sum = x_orientation(0) * y_orientation(0) + x_orientation(1) * y_orientation(1) +
+             x_orientation(2) * y_orientation(2);
+
+  if (std::abs(sum) > 1e-4)
+    mooseError("ComputeIncrementalBeamStrain: y_orientation should be perpendicular to "
+               "the axis of the beam.");
+
+  // Calculate z orientation as a cross product of the x and y orientations
+  RealGradient z_orientation(3, 0.0);
+  z_orientation(0) = (x_orientation(1) * y_orientation(2) - x_orientation(2) * y_orientation(1));
+  z_orientation(1) = (x_orientation(2) * y_orientation(0) - x_orientation(0) * y_orientation(2));
+  z_orientation(2) = (x_orientation(0) * y_orientation(1) - x_orientation(1) * y_orientation(0));
+
+  // Rotation matrix from global to original beam local configuration
+  _original_local_config[_qp](0, 0) = x_orientation(0);
+  _original_local_config[_qp](0, 1) = x_orientation(1);
+  _original_local_config[_qp](0, 2) = x_orientation(2);
+  _original_local_config[_qp](1, 0) = y_orientation(0);
+  _original_local_config[_qp](1, 1) = y_orientation(1);
+  _original_local_config[_qp](1, 2) = y_orientation(2);
+  _original_local_config[_qp](2, 0) = z_orientation(0);
+  _original_local_config[_qp](2, 1) = z_orientation(1);
+  _original_local_config[_qp](2, 2) = z_orientation(2);
+
+  // For small rotation problems, the rotation matrix is essentially the transformation from the
+  // global to original beam local configuration and is never updated
+  _total_rotation[_qp] = _original_local_config[_qp];
+
+  _K11[0].zero();
```

Comment:
A lot of non-stateful properties are being set here, which is a problem because they'll be lost.